server/bin/pbench-dispatch.sh

```diff
@@ -185,6 +196,28 @@ while read tbmd5; do
         continue
     fi
 
+    if [[ ! -z "${put_token}" ]]; then
+        # We have a bearer token for the PUT API of a "new" Pbench Server,
+        # invoke the `pbench-results-push` agent CLI to send it to the
+        # server (configuration of that server handled elsewhere).
+
+        # All tar balls pushed to the "new" Pbench Server are made public just
+        # like they are all public on the current "old" Pbench Server.
+        push_options="--token ${put_token} --access=public"
+        push_options+=" --metadata=server.legacy.version:${server_version}"
+        push_options+=" --metadata=server.legacy.sha1:${server_sha1}"
+        push_options+=" --metadata=server.legacy.hostname:${server_hostname}"
```

Comment:
This is cute, but it won't actually work as `server` is a controlled metadata namespace. Only `server.deletion` and `server.origin` can be set. You could try something more like this, though:

```
    origin="version:${server_version}/sha1:${server_sha1}/host:${server_hostname}"
    if [[ "${controller}" != "${satellite}" ]]; then
        origin+="/satellite:${satellite}"
    push_options+=" --metadata=server.origin:${origin}"
```

(And, yes, the key/value is split using `split(":", maxsplit-1)`, so the additional ":" characters won't cause a problem, but you can't use `,`; and while `=` isn't a problem for *our* code, I'm not sure about Flask's query parameter parsing.)

That would be searchable with, e.g., `server.origin:~/satellite:RIYA`.

Another option would be to use `global.legacy` as an ad-hoc legacy compatibility sub-namespace, where the code you've got would otherwise work as-is. (And `GET /datasets/metadata/{id}?metadata=global.legacy` would return the `version`, `sha1`, `hostname` and `satellite` as a JSON object, which would be neat packaging.)

A third option would be server changes to allow `server.origin` to behave as a JSON sub-document, so you could set `server.origin.version` and such: and, again, retrieving `server.origin` would give a neat JSON object. I actually considered that when I implemented it, but "kept it simple". 👀

Reply:
Okay, that is a bummer.  It'd be great to have the abiility to set arbitrary `server` namespace data if we have the permission to do so instead of having the server enforce a restricted name space.

Reply:
There is a `global` namespace available for setting the arbitrary data, but yeah I don't think `global` is more appropriate here for setting the hostname origin.

Reply:
> Yeah, that's exactly the problem "if we have the permission". How do we establish that? In theory, with Keycloak, we can give the put-shim server's account a special trusted role, but right now there's nothing special about it.

Not sure what you mean.  We have a bearer token.  That token should be given out by the Pbench Server to allow the use of the PUT API with the ability to set metadata in the `server` namespace for this application.

How that is implemented on the server is TBD, but ideally this code should assume that it can do that.

I am using the `global.server` namespace because I don't have any other mechanism.  I don't think we want to wait for a change on the new Pbench Server side, and mushing all the metadata into one string to overload the value of `server.origin` doesn't seem like a good approach either because it doesn't make the data that much more useful or accessible.

If somebody feels so inclined to implement a `server` namespace that is arbitrary for authorized users  before our mid-March deadline, that would be great.  Otherwise, this will do for now.

Reply:
> There is a `global` namespace available for setting the arbitrary data, but yeah I don't think `global` is more appropriate here for setting the hostname origin.

Can you propose something that is more appropriate?